masterslave.q

```diff
@@ -1,62 +1,88 @@
 //Master slave algorithm
 //Each process initially assumes it's the master process
-//Master process is determined by oldest process start time 
+//Master process is determined by oldest process start time
 //Connect to other slave processes and retrieve details to determine which process is master
 
 //run lines
 //q torq.q -load masterslave.q -proctype lb -procname lb1 -p 6100 -debug -parentproctype wdb
 //q torq.q -load masterslave.q -proctype lb -procname lb2 -p 6101 -debug -parentproctype wdb
 
-
-// the list of processes to connect to
-.servers.CONNECTIONS:`lb
-
-// custom function, this is invoked when a new outbound connection is created
-// to be customised to invoke negotiation of processes
-.servers.connectcustom:{.lg.o[`connect;"found new connection"]; show x}
-
-// create connections
-.servers.startup[]
+//test cases
+//one process starts up, it becomes master
+//two processes, only one is master
+//two processes, master dies, slave takes over
+//two processes, one is made master, new process added, original master(if it's starttime
+// is still oldest)
+//two processes, one is made master, an older process is connected, recently connected process
+// should then become master
+//3 processes, one master, master dies, one of the remaining two takes over
+
+init:{
+ // the list of processes to connect to
+ .servers.CONNECTIONS:distinct .servers.CONNECTIONS,.proc.proctype;
+
+ // custom function, this is invoked when a new outbound connection is created
+ // to be customised to invoke negotiation of processes
+ .servers.connectcustom:{.lg.o[`connect;"found new connection"];
+                         .masterslave.checkifmaster[];
```

Comment:
If we get a new connection in, then we only want to get the details for that single new connection. WE don't want to re-run the whole thing for all connections, which I think is what is happening here.